### DISCUSSION TITLE
Problem in joining multiple rooms

### DISCUSSION THREAD

**Discussion by satyam73** [socketio/socket.io#4340] [Q&A]
Hello all,
I am creating a chat app with authentication. I used jwt token and cookies authentication. I want that when I click on certain user the room will get created with same room id for both the use which is stored in both user1 and user2's database. I added this feature when user clicks on any user then emit join chat event and then join the room. But when I click on user1 it works well and when I click on user2 it doesn't join another room which should be joined.

Example : Let's say we have three user user1,user2, and user3. User1 has one contact which is user2, user3 has one contact which is user2, and user3 has two contact which is user1 and user2.

case 1 : so when i open three browser tabs and opens all the three account separately and when from user1 account I click user2 and from user2 I click user1. So in this case all things works perfectly. 

case 2: But now since user 2 has two connected people and I want to chat with user3 so when I click on user3 from user2's account and from user3 i click on user2 account , nothing works and when I get back to the case 1 where user1 and 2 are chatting now that doesnt work after this and when I reload the page and follow the case1 so the case1 works.

That is user2 is not joining other room after I click on user3.

**Comment by darrachequesne** [comment]
Hi! That sounds weird. Could you please provide an example reproducing the issue?

You can start from the basic project here: https://github.com/socketio/socket.io-fiddle

**Reply by satyam73** [reply]
This is the onclick function where I am emitting the event.

```
async function chattingWith(elem) {
    try {
        remove();
        let message = chatInput.value
        elem.classList.add("chatting");

        selectedUser();
        let sender = await getUserHandler();
        console.log('chatting with function fires ', userSelected);
        console.log(sender);
        console.log(userSelected)
        let id;
        chatID = makeid(20)
        populateChats(message);
        // console.log(obj);
        sender.contacts.forEach((element) => {
            if (!element.chatID) {
                let obj = { chatID, toUser: userSelected, sender: sender._id };
                fetch("http://localhost:8080/api/setChatID", {
                    method: 'POST',
                    body: JSON.stringify(obj),
                    headers: {
                        'Content-Type': 'application/json'
                    }
                }).then(res => {
                    return res.json();
                }).then(data => {
                    console.log(data);
                }).catch(err => {
                    console.log(err);
                })
            }
        })

        sender.contacts.forEach((element) => {
            if (userSelected == element.contactID) {
                console.log(true);
                if (element.chatID) {
                    id = element.chatID

                    console.log('chatID', id);
                    // socket.emit("join chat", id);

                }
            }
        })
        await checkAuth();
   
        socket.emit("join chat", id);

        console.log('chattingwith : function completed');
    } catch (err) {
        console.log(err);
    }
}
```

This is the socket function in app js

```
// const newMessage = "hello world"
const socketIO = (io) => {
    io.on("connection", socket => {
        console.log("connected to socket");

        socket.on("setup", (user) => {
            socket.leaveAll()
            socket.join(user);
            console.log("setup ", user);
            socket.emit("connected");
        })
        socket.on("join chat", room => {
            socket.leaveAll();
            console.log(io.sockets.adapter.rooms);
            socket.join(room);
            console.log("user joined room ", room);
            console.log(io.sockets.adapter.rooms);
            // socket.broadcast.to(room).emit("invite", room)
        })
        // socket.on('joinRoom', (room) => {
        //     socket.join(room);
        // });
        socket.on("new message", (newMessage) => {
            // console.log(io.rooms);
            if (!newMessage.to || !newMessage.from) { return console.log("chat uses not defined"); }

            // if (newMessage.from == req.user_id) { return }
            socket.in(newMessage.chatID).emit("message received", newMessage)
            console.log(newMessage.chatID);
            console.log(`the new message is ${newMessage.message}`);
        });


    })
}

module.exports = socketIO;
```

**Reply by satyam73** [reply]
This chattingWith function triggers when I click on any user. When I click on any user first time it joins the room, but when I click again on another user it doesn't join the room whereas the function triggers.

**Reply by darrachequesne** [reply]
I think this might be due to the usage of `socket.leaveAll()`. This is a private method that is not meant to be used by end users.

Could you please check?
